text/0000-at-least-one-feature.md

```diff
@@ -0,0 +1,293 @@
+We still have the restriction that new features can be added below, but this restriction is no longer a problem:
+there is no point of adding such a new minimal feature because there is nothing left to `cfg`-out!
+
+This solution is more mathematically elegant, but it seems harder to implement.
+It is unclear how Cargo could require the Rust code to obey this property without new infra like the portability lint.
+
+Finally, in a sense these variations are actually isomorphic.
+If no one can request the empty feature set, no one can observe what library interface that corresponds to.
+There is no reason to conclude that it *isn't* the empty library.
+
+## RFC [#3146](https://github.com/rust-lang/rfcs/pull/3146)
+
+This more more expensive solution aimed to solve the "running out of good names" problem this has.
+
+Returning to our commutative diagram, instead of
+
+- removing the ⊥ feature downset (this proposal)
+
+- mandating the ⊥ feature downset correspond to ⊥ the empty library interface (first alternative, extra rules on the vertical mappings in diagrams)
+
+We can instead
+
+- allow customizing the mapping from old feature names to new feature names (customizing top horizontal mapping)
+
+This works and allows giving things the "right" names" every time, but is much work work to implement.
+
+I think it might make sense to do this someday, but it is better to start with this.
+My thinking about this is basically "walk before run".
+Today, that would be lots of work for an obscure problem, and a solution that downstream crates need to know about because it affects how dependencies are written.
+However if we do this first, "opt-out features" are poised to become much more popular now that they are no longer a compatibility hazard.
+We can evaluate both that new popularity, and the incidence of the "bad name" problem in *practice*, and use that to justify a more heavyweight solutions.
+
+## RFC [#3283](https://github.com/rust-lang/rfcs/pull/3283)
+
+This proposal adds a new notion of a "feature base", which aims to essential provide more `default-features = false` options rather than just one forever.
+`default-features = false` is ret-conned as the first feature base.
+
+My view is that that RFC is pretty close to being this one in disguise.
+A "feature base" is just a feature.
+The fact that there is always one "feature base" is just another way of accomplishing the "at least one feature"
+restriction.
+
+The main difference is that `default-features = false` is ret-conned as a feature to allow for existing users.
+That can be thought of from the perspective of this RFC as a "one-off" feature migration to get reverse dependencies predating the use of `at-least-one-feature = true` on to a feature so they retroactively abide by the "at least one feature" rule.
```

Comment:
@tbu- I think you might be on to something :)

Prioritizing this single feature migration before the others complete side-steps the "is `at-least-one-feature = true` a breaking change?" problem, by giving `default-features = false` users an on-ramp to the brave new world.

What do you think @epage? If anyone wants to throw out some syntax I am happy to make this part the RFC. There could be a `mini-default` feature that handles the `default-features = false` for example, `default`'s little sibling.

@epage in https://github.com/rust-lang/cargo/issues/3126 proposed deprecating `default-features = <bool>` and I think that is a good idea too. I will expand the "future work" section" to discuss this.

Reply:
The problem with `default-features = false` implying a feature when `[]` but not when `["base"]` is that those are very unclear / confusing semantics.

However, I think we can combine this proposal with negative features and come up with something that might resolve problems with both.

First, we need to step back and talk about the two main uses for `default-features = false`
- As a clap user, there are a lot of default policies to provide a good experience.  People usually remove them piece meal in their applications (negative features)
- As the author of `clap-verbosity-flag`, I do not want to enable any clap features but the minimum (`default-features = false`)

Negative features helps with one, this proposal helps with another.

What if we mixed
- `package.at-least-one-feature = true`
- `default-features = false` migrations
- Negative features

The pieces
- Add `package.empty-features` field, defaulted to `true`
  - Yeah, I'm mixing in bike shedding over the name
  - `cargo new` will write populate the manifest with `package.empty-features = false`
  - On the next edition, we change the default to `package.empty-features = false`
- Allow soft requirements on features: the feature ends with `?`.
  - Features required with a `?` can be removed **when** resolving the list of features associated with a specific dependency.  Much like `default-features = false`, this does not conflict with other dependencies that use the default feature
  - Dependent features without `?` are hard errors (or maybe disabled?)
    - Errors allow adding `?` without a breaking change
  - Note:  you can remove features from anywhere, not just `default`
- On a new edition, we remove `default-features` field in favor of `default` feature
  - We treat `no-default-features` as special feature for a onetime migration from `package.empty-features = true` to `package.empty-features = false` without a breaking change
  - All features that ` no-default-features` requires will have to use `?`
  - Note: we'd need to check if `no-default-features` is safe to add semantic meaning to
  - Command-line behavior is unchanged

We now have a way to add the `http2` feature to warp without adding a bunch of `without-http2` features

We can migrate to this without a breaking change.  Note that I consider this an optional requirement.  So long as we move people to using this for new crates and enough crates are not 1.0-forever, we might be able to get away with a breaking change.

People can choose to disable everything or disable piece meal.

I expect there are still problems with this but I'm intrigued with the vague idea of aligning these proposals.

Reply:
> The problem with `default-features = false` implying a feature when `[]` but not when `["base"]` is that those are very unclear / confusing semantics.

Agreed. In fact it is downright wrong, because old `["base"]` feature lists still also need the migration. I was not clear but any one-off attempt at `default-features = false` migrations will also need to distinguish between old and new dependencies.

Still reading the rest.

Reply:
> Yeah, I'm mixing in bike shedding over the name

Sure, I don't care about the name. I'll just go change it.

Reply:
@epage So I agree that it might possibly be convenient for end executable offers to opt-out of a few clap features. But I still l find it very degenerate for reusable libraries to express anti-dependencies. I think the only proper way to write a reusable library is to stick to purely additive features --- *say what you need, don't waste text on what you don't need*.

I think the `clap-verbosity-flag` use-case corresponds to my "reusable library" use-case? But the `foo-without-http2` problem effects reusable libraries too.

Thus when you write

> Negative features helps with one, this proposal helps with another.

I would like that but it doesn't seem quite true? We have two solutions and two use-cases, but they don't work in isolation.

Reply:
My discussion on the use cases (whether to focus on negative features or disabling default features) is on the dependent side (a binary using clap, a library that integrates with clap).  These would be starting points for thinking.  This does not mean they are mutually exclusive, two features tacked on for different cases, but two intertwined features that overall enhance each other.   Like I said, `warp` can add its `http2` feature and an warp-integration library might use feature negation but that is fine.

To put it another way, the feature migration feature would be moving away from specifically trying to solve default features to being a general feature that also enhances this one, removing rough edges.

Some of this idea of `?` feature requirements stems from @djc's [comments about gentoo](https://github.com/rust-lang/cargo/issues/3126#issuecomment-1299959611) though I don't know if the exact idea I mentioned exists in gentoo.

Reply:
The `?` mentioned there is a bit unclear, but it is still "additive". It seems to boil down to adding extra dependencies to conjunctions of features (enabling foo and `foo` and `bar` also enables `baz`). I don't see what that has to do with negative dependencies.

> Like I said, `warp` can add its `http2`

But that only works without the "odd named features" if we *force* all downstream deps to use negative feature "requests" to avoid `http2`. If I am writing a downstream library, a nice nugget not a big "integrationy" one for sake of argument, I don't want to use negative dependencies. I want to just regularly "additively" depend on just the parts of `warp` I need, and I don't want to waste any mental effort or keystrokes on features I don't care.

I don't see two intertwined features, I see distasteful negative reasoning as a cure for odd feature names that is worse than the disease. If some other use-case wants negative features I would still think they are mistaken (X Y problem), but moreover I see it bleeding into this problem when it should be orthogonal.